src/network.rs

```diff
@@ -44,52 +43,61 @@ impl Network {
             // TODO: make this configurable on the CLI or something
             let timeout = Duration::from_secs(DEFAULT_TIMEOUT_SECS);
             // TODO: make this configurable on the CLI or something
-            let in_progress_download_dir = std::env::temp_dir();
             let client = Client::builder()
                 .timeout(timeout)
                 .build()
                 .expect("Couldn't construct HTTP Client?");
             Some(Arc::new(Self {
                 client,
-                in_progress_download_dir,
+                connection_semaphore: tokio::sync::Semaphore::new(MAX_CONCURRENT_CONNECTIONS),
             }))
         }
     }
 
     /// Download a file and persist it to disk
-    pub async fn download_and_persist(
-        &self,
-        url: Url,
-        persist_to: &Path,
-    ) -> Result<File, VetError> {
-        let mut res = self
-            .client
-            .get(url.clone())
-            .send()
-            .await
-            .and_then(|res| res.error_for_status())
-            .wrap_err_with(|| format!("Failed to download {}", url))?;
+    pub async fn download_and_persist(&self, url: Url, persist_to: &Path) -> Result<(), VetError> {
+        let download_tmp_path = OsString::from_iter([persist_to.as_os_str(), OsStr::new(".part")]);
+        {
+            let _permit = self.connection_semaphore.acquire().await?;
```

Comment:
interesting, we want to limit network ops but not disk ops? (genuine question, curious)

Reply:
We need to limit network operations because otherwise we'll get network errors because crates.io will stop accepting new network connections from us. I could actually run into this on mozilla-central when fetching crates.

Technically I think we might need to do even better than this (there's an official crawler policy for crates.io, but I don't think cargo-vet will do bad enough to run into those hard rate limits: https://crates.io/policies#crawlers)

I didn't bother with the file I/O. That's already somewhat limited by the number of blocking threads in the tokio runtime (the default appears to be 512: https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.max_blocking_threads), and it doesn't seem to be causing us any issues. We're limited enough by the network that file I/O being a problem doesn't come up, and we want to try to go as fast as we can with that bottleneck I'd think.

If in the future it comes up as an issue, we can look into a higher limit for other types of I/O.